metric_learn/pgdm.py

```diff
@@ -0,0 +1,436 @@
+"""
+Probabilistic Global Distance Metric Learning, Xing et al., NIPS 2002
+
+PGDM minimizes the sum of squared distances between similar examples,
+while enforcing the sum of distances between dissimilar examples to be
+greater than a certain margin.
+This leads to a convex and, thus, local-minima-free optimization problem
+that can be solved efficiently.
+However, the algorithm involves the computation of eigenvalues, which is the
+main speed-bottleneck.
+Since it has initially been designed for clustering applications, one of the
+implicit assumptions of PGDM is that all classes form a compact set, i.e.,
+follow a unimodal distribution, which restricts the possible use-cases of
+this method. However, it is one of the earliest and a still often cited technique.
+
+Adapted from Matlab code at http://www.cs.cmu.edu/%7Eepxing/papers/Old_papers/code_Metric_online.tar.gz
+"""
+
+from __future__ import print_function, absolute_import
+import numpy as np
+from six.moves import xrange
+from sklearn.metrics import pairwise_distances
+from sklearn.utils.validation import check_array, check_X_y
+
+from .base_metric import BaseMetricLearner
+from .constraints import Constraints
+
+
+# hack around lack of axis kwarg in older numpy versions
```

Comment:
This code is used in ITML as well, so let's pull this hack out into a new file, `metric_learn/_util.py`.